src/layout/tree.rs

```diff
@@ -0,0 +1,194 @@
+        // Make all the views in the original workspace to be invisible
+        for view in old_workspace.get_children_mut() {
+            trace!("Setting {:?} invisible", view);
+            match view.get_val().get_handle().unwrap() {
+                Handle::View(view) => view.set_mask(0),
+                _ => {},
+            }
+        }
+    }
+    let current_workspace: *const Node;
+    {
+        let new_current_workspace: &mut Node;
+        if let Some(_) = get_workspace_by_name(&tree, name) {
+            trace!("Found workspace {}", name);
+            new_current_workspace = get_workspace_by_name_mut(&mut tree, name).unwrap();
+        } else {
+            drop(tree);
+            try!(add_workspace(name));
+            tree = try!(TREE.lock());
+            new_current_workspace = get_workspace_by_name_mut(&mut tree, name).unwrap();
+        }
+        for view in new_current_workspace.get_children_mut() {
+            trace!("Setting {:?} visible", view);
+            match view.get_val().get_handle().unwrap() {
+                Handle::View(view) => view.set_mask(1),
+                _ => {},
+            }
+        }
+        // Set the first view to be focused, so that the view is updated to this new workspace
+        if new_current_workspace.get_children().len() > 0 {
+            trace!("Focusing view");
+            match new_current_workspace.get_children_mut()[0].get_val().get_handle().unwrap() {
+                Handle::View(view) => view.focus(),
+                _ => {},
+            }
+        } else {
+            WlcView::root().focus();
+        }
+        current_workspace = new_current_workspace as *const Node;
+    }
+    tree.active_container = current_workspace;
+    Ok(())
+}
+
+/// Finds the WlcOutput associated with the WlcView from the tree
+pub fn get_output_of_view(wlc_view: &WlcView) -> Option<WlcOutput> {
+    let tree = TREE.lock().unwrap();
+    if let Some(view_node) = tree.root.find_view_by_handle(wlc_view) {
+        if let Some(output_node) = view_node.get_ancestor_of_type(ContainerType::Output) {
+            if let Some(handle) =  output_node.get_val().get_handle() {
+                return match handle {
+                    Handle::Output(output) => Some(output),
+                    _ => None
+                }
+            }
+        }
+    }
+    return None;
+}
+
+fn get_focused_workspace<'a>(tree: &'a Tree) -> Option<&'a Node> {
+    for output in tree.root.get_children() {
+        if output.get_val().is_focused() {
+            for workspace in output.get_children() {
+                if workspace.get_val().is_focused() {
+                    return Some(workspace);
+                }
+            }
+        }
+    }
+    None
+
+}
+
+fn get_current_workspace<'a>(tree: &'a Tree) -> Option<&'a mut Node> {
+    if let Some(container) = tree.get_active_container() {
+        //if let Some(child) = container.get_ancestor_of_type(ContainerType::Workspace) {
```

Comment:
The commented out code _does_ work **if** get_active_container actually did what it did, right now I've hacked it so it just returns the active workspace, so that commented out code would never work (since we that code is trying to find that workspace). I should have fixed that up so it's easier to see that, sorry. I can get to that later tonight, though feel free to rename things to make more sense if you'd like/can